driver: dma: dma_silabs_siwx91x: Add pm device support for dma driver

fimohame opened the pull request:
This commit enables the pm device driver support for the dma_silabs_siwx91x driver.

jerome-pouiller commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -591,7 +633,7 @@ static int siwx91x_dma_init(const struct device *dev)
 	}
 
-	return 0;
+	return pm_device_driver_init(dev, dma_siwx91x_pm_action);
```
`pm_device_driver_init()` is going to call `dma_siwx91x_pm_action(dev, PM_DEVICE_ACTION_RESUME)`. So, you can factorize `dma_siwx91x_init()` and `dma_siwx91x_pm_action()`.

fimohame replied:
It is required to have them!

fimohame committed:
driver: dma: dma_silabs_siwx91x: Add pm policy state support for dma driver

This commit enables the pm policy state lock support
for the dma_silabs_siwx91x driver.

Signed-off-by: S Mohamed Fiaz <fiaz.mohamed@silabs.com>

jerome-pouiller commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -573,27 +573,39 @@ static int siwx91x_dma_init(const struct device *dev)
-		return -EINVAL;
-	}
+		udma_handle = UDMAx_Initialize(&udma_resources, udma_resources.desc, NULL,
+					       (uint32_t *)&data->udma_handle);
```
Do you think `UDMAx_Initialize()` should be called on `PM_DEVICE_ACTION_RESUME` or on `PM_DEVICE_ACTION_TURN_ON`?

I believe Device runtime PM will only play with `PM_DEVICE_ACTION_SUSPEND`. `PM_DEVICE_ACTION_SUSPEND` may stop the clocks and the regulator but I believe it will retains the registers.

I assume System Managed PM won't retains the registers but it will call `PM_DEVICE_ACTION_TURN_OFF` and `PM_DEVICE_ACTION_TURN_ON`. So, `UDMAx_Initialize()` should be only done in `PM_DEVICE_ACTION_TURN_ON`.

@Martinhoff-maker, @asmellby, can you confirm my understanding?

fimohame commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -573,27 +573,39 @@ static int siwx91x_dma_init(const struct device *dev)
-	}
+	if (action == PM_DEVICE_ACTION_RESUME) {
+		ret = clock_control_on(cfg->clock_dev, cfg->clock_subsys);
+		if (ret) {
```
addressed

fimohame commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -573,27 +573,39 @@ static int siwx91x_dma_init(const struct device *dev)
+			return -EBUSY;
+		}
+	} else if (IS_ENABLED(CONFIG_PM_DEVICE) && (action == PM_DEVICE_ACTION_SUSPEND)) {
+		return 0;
```
Ain't sure if its required for DMA

Martinhoff-maker reviewed (commented):
Just to make sure that we got the same understanding (and it's between all the PM PR):

Since you got no runtime get and put inside of this uDMA driver, you consider that the device will always be suspended or turned off if the power domain is suspended ? Then since we are on the same power-domain than the peripheral that use the uDMA, when you will do a "get" inside of this peripheral, it will turn on the power domain and then put the uDMA from OFF to suspended (but that will be good since everything's is alive in our "suspended" mode) ?

fimohame commented on dts/arm/silabs/siwg917.dtsi:
```diff
@@ -304,6 +304,8 @@
 			#dma-cells = <1>;
 			dma-channels = <32>;
+			power-domains = <&siwx91x_soc_pd>;
+			zephyr,pm-device-runtime-auto;
```
addressed

fimohame commented:
I have updated with pm policy state lock get and put for the driver.

Martinhoff-maker commented:
Edit: After discussing with @asmellby, I think that using pm_policy lock is mandatory. that's ok. how do you test that it works properly ?

fimohame commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -60,6 +62,18 @@ struct dma_siwx91x_data {
 					      */
 };
 
+static void siwx91x_dma_pm_policy_state_lock_get(const struct device *dev)
```
addressed

jerome-pouiller commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -60,6 +63,18 @@ struct dma_siwx91x_data {
+static void siwx91x_dma_pm_policy_state_lock_get(void)
+{
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	pm_policy_state_lock_get(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
```
The dtsi declare these power states:
```
               power-state-name = "runtime-idle";
               power-state-name = "suspend-to-idle";
```
So, `PM_STATE_STANDBY` -> `PM_STATE_RUNTIME_IDLE`.
